src/prefect/cache_policies.py

```diff
@@ -157,6 +157,40 @@ class CompoundCachePolicy(CachePolicy):
 
     policies: list[CachePolicy] = field(default_factory=list)
 
+    def __sub__(self, other: str) -> CachePolicy:
+        if not isinstance(other, str):  # type: ignore[reportUnnecessaryIsInstance]
+            raise TypeError("Can only subtract strings from key policies.")
+
+        # Subtract from each sub-policy; this may stack new Inputs(...) layers
+        new_subpolicies: list[CachePolicy] = [p - other for p in self.policies]
```

Comment:
@zzstoatzz
`p - other` can create a new CompoundCachePolicy with nested Inputs.
This can cause unintended layering of Inputs within CompoundCachePolicy.
You might need to recursively update the `exclude` field for nested Inputs as well.

Reply:
It also seems that with the standard implementation of __sub__, this approach tends to lead to exponential growth in complexity due to nested CompoundCachePolicy objects.

Reply:
hi @a14e - the initial draft was updated via #16786 - feel free to take a look now!

Reply:
Hi @zzstoatzz,  
The new implementation is much better!

However, I think there might be an issue with this case:  
```python
(RunId() + (TaskSource() - "key1")) - "key2"
```

It would generate an object like this:  
```python
CompoundCachePolicy(
    policies=[
        RunId(),
        CompoundCachePolicy(policies=[TaskSource(), Inputs(exclude=["key1"])]),
        Inputs(exclude=["key2"])
    ]
)
```

This means the inner `Inputs` won’t know anything about `"key2"`.

Perhaps it makes sense to move the unpacking of `CompoundCachePolicy` into `__post_init__`?

Reply:
Good call @a14e! The flattening of `CompoundCachePolicy` instances should be handled by https://github.com/PrefectHQ/prefect/pull/16785/commits/c42d0552520a89b1b96378dbafc95408c3edc5c4.

Reply:
@desertaxle looks great